### PR TITLE
Add org-modern-custom-timestamp-format

### PR DESCRIPTION
**Pull request by vizs**
Continuation of #29.  I tried to cleanup the docstring of the new option.  I believe I exhausted the 15 LOC since I had a merge patched in Emacs.  I mailed assign@gnu.org and I am currently waiting for their reply.

### CODE REVIEW

File: org-modern.el [hunk 1]
```diff
@@ -92,6 +92,13 @@ Set to nil to disable styling the headlines."
 Set to nil to disable styling the time stamps."
   :type 'boolean)
 
+(defcustom org-modern-custom-timestamp-format '("%Y-%m-%d" . "%H:%M")
```
Comment (minad) [inline org-modern.el:95]: Set this to nil by default.

File: org-modern.el [hunk 2]
```diff
@@ -350,16 +365,28 @@ You can specify a font `:family'. The font families `Iosevka', `Hack' and
      (match-beginning 0)
      (if (eq (match-beginning 2) (match-end 2)) (match-end 0) (match-end 1))
      'face date-face)
+    (when org-display-custom-times
```
Comment (minad) [inline org-modern.el:368]: Check the variable org-modern-custom-time-format instead if it is nil.

File: org-modern.el [hunk 3]
```diff
@@ -511,7 +538,7 @@ You can specify a font `:family'. The font families `Iosevka', `Hack' and
            (2 'org-modern-block-keyword append))))
       (when org-modern-tag
         '(("^\\*+.*?\\( \\)\\(:.*:\\)[ \t]*$" (0 (org-modern--tag)))))
-      (when (and org-modern-timestamp (not org-display-custom-times))
+      (when org-modern-timestamp
         '(("\\(?:<\\|\\[\\)\\([0-9]\\{4\\}-[0-9]\\{2\\}-[0-9]\\{2\\}\\(?: [[:word:]]+\\)?\\(?: [.+-]+[0-9ymwdh/]+\\)*\\)\\(\\(?: [0-9:-]+\\)?\\(?: [.+-]+[0-9ymwdh/]+\\)*\\)\\(?:>\\|\\]\\)"
            (0 (org-modern--timestamp)))
```
Comment (minad) [inline org-modern.el:543]: Instead call a separate function here for the custom time format. We will have org-modern--timestamp-default and org-modern--timestamp-custom.

**Review by minad — CHANGES_REQUESTED** [review]

File: org-modern.el (hunk 1 above)
Comment (vizs) [reply org-modern.el]: Done

File: org-modern.el [hunk 4]
```diff
@@ -361,6 +371,53 @@ You can specify a font `:family'. The font families `Iosevka', `Hack' and
        (match-end 0)
        'face time-face))))
 
+(defun org-modern--timestamp-custom ()
```
Comment (minad) [inline org-modern.el:374]: Thanks! I hope the code can be simplified a bit.

File: org-modern.el [hunk 5]
```diff
@@ -361,6 +371,53 @@ You can specify a font `:family'. The font families `Iosevka', `Hack' and
...
+                      'org-modern-date-inactive))
+         (time-face (if active
+                        'org-modern-time-active
+                      'org-modern-time-inactive))
+         (time (save-match-data
+                 (encode-time
+                  (org-fix-decoded-time
+                   (org-parse-time-string
+                    (buffer-substring (match-beginning 0) (match-end 0))))))))
+    (remove-text-properties (match-beginning 0) (match-end 0) '(display nil))
```
Comment (minad) [inline org-modern.el:388]: `remove-list-of-text-properties`

File: org-modern.el [hunk 6]
```diff
@@ -361,6 +371,53 @@ You can specify a font `:family'. The font families `Iosevka', `Hack' and
...
+    (put-text-property
+     (match-beginning 0)
+     (1+ (match-beginning 0))
+     'display " ")
```
Comment (minad) [inline org-modern.el:392]: Remove. Make the space part of org-odern-custom-timestamp-format.

File: org-modern.el [hunk 7]
```diff
@@ -361,6 +371,53 @@ You can specify a font `:family'. The font families `Iosevka', `Hack' and
...
+    (put-text-property
+     (1- (match-end 0))
+     (match-end 0)
+     'display " ")
```
Comment (minad) [inline org-modern.el:396]: Remove.

File: org-modern.el [hunk 8]
```diff
@@ -361,6 +371,53 @@ You can specify a font `:family'. The font families `Iosevka', `Hack' and
...
+    ;; year-month-day
+    (put-text-property
+     (match-beginning 0)
+     (if (eq (match-beginning 2) (match-end 2)) (match-end 0) (match-end 1))
+     'face date-face)
+    (put-text-property
+     (match-beginning 1)
+     (match-end 1)
+     'display (format-time-string (car org-modern-custom-timestamp-format) time))
```
Comment (minad) [inline org-modern.el:405]: Probably these two calls to put-text-property can be merged into one add-text-properties call.

File: org-modern.el [hunk 9]
```diff
@@ -361,6 +371,53 @@ You can specify a font `:family'. The font families `Iosevka', `Hack' and
...
+       (match-end 1)
+       'display " ")
+      (put-text-property
+       (match-beginning 2)
+       (match-end 0)
+       'face time-face)
+      (put-text-property
+       (match-beginning 2)
+       (match-end 2)
+       'display (concat " " (format-time-string (cdr org-modern-custom-timestamp-format) time))))))
```
Comment (minad) [inline org-modern.el:419]: Same here. These calls can probably be merged too.

**Comment by minad** [comment]
Please take your time to polish this a little bit. There is no hurry and we have to wait for the FSF assignment anyway.

File: org-modern.el (hunk 5 above)
Comment (vizs) [reply org-modern.el:388]: Okay, I wasn't sure which function to prefer.

File: org-modern.el (hunk 6 above)
Comment (vizs) [reply org-modern.el:392]: Then I suppose I need to document this part i.e., that we need an explicit space at the end of date format?

Comment (minad) [reply org-modern.el:392]: Yes this should be documented. You can also give example configurations as part of the :type specification of the defcustom.

**Comment by minad** [comment]
@vizs What is the status here?

**Comment by vizs** [comment]
I'm currently in the middle of exam season but even if I manage to find the time to finish this up, we still have to wait for FSF to reply.  They haven't sent back me the form (it's been a month)... I mailed copyright-clerk@gnu.org two days before for a status update.

**Comment by minad** [comment]
Okay please notify me!

**Comment by vizs** [comment]
My copyright assignment is finished now.  I hope to get to the reviews on 29th or 30th Apr.  Thanks for your patience

**Comment by vizs** [comment]
@minad Please check if everything is okay

**Comment by vizs** [comment]
Oops, looks like there are some fallouts after the merge (I did not check the relevant hunk carefully!), I will get to it this evening.

**Comment by minad** [comment]
Okay, I merged your PR. I went over it again and thought that we could compress things a little more by simply reusing the org-modern-timestamp variable also for custom formats, see https://github.com/minad/org-modern/commit/ff1046705b3950b7a49da50bc34c11da86c6226d. Please check if everything still works okay for you.

**Pull request merged**